### PR TITLE
Database migration

### PR DESCRIPTION
**Pull request by BattleMage0231**

### CODE REVIEW

File: application.py
Comment (jdabtieu) [inline application.py:391]: misspelling: should be contest_problems, not context_problems

**Comment by jdabtieu** [comment]
Please undraft when you are ready to merge. Thank you!

File: application.py
Comment (BattleMage0231) [reply application.py]: I believe that's fixed in the latest commit. I'm still working on fixing bugs since a lot were generated because of the change.

**Comment by BattleMage0231** [comment]
I believe the build fails because of outdated database tests. That can be fixed in a separate issue (as well as possibly changing migrate.py) as database migration seems to be holding back a lot of other issues.

**Comment by BattleMage0231** [comment]
There may still be some bugs as this is a somewhat big change. However, I've fixed all the obvious ones and contests & submission appears to work fine locally.

Solves #28.

**Comment by jdabtieu** [comment]
Yep, as I expected, some of the database-related tests would fail due to an outdated format. I will manually review the code later.

@ComputerGenius152 please wait for my review to be approved before merging, this is quite a large change and I have noticed some preliminary bugs to be fixed first.

**Comment by ComputerGenius152** [comment]
I think the test failures aren't database-related. It's because maintenance mode is never disabled at the end of the maintenance test so all later tests fail.

File: application.py
Comment (ComputerGenius152) [inline application.py:75]: this section is causing test_maintenance and later test to fail, can you take a look

**Review by ComputerGenius152 — CHANGES_REQUESTED** [review]

File: application.py
Comment (BattleMage0231) [reply application.py:75]: That section was created during resolution of merge conflicts. I thought it was already like that in master?

File: templates/problem/problems.html
Comment (jdabtieu) [inline templates/problem/problems.html:20]: Initial review: looks good

**Comment by BattleMage0231** [comment]
The previous error should be resolved. The one remaining seems to be an error with tests as they don't create the tables properly anymore.

**Review by jdabtieu — CHANGES_REQUESTED** [review]
Initial review: just looking at the code and seeing how it functions
I will be doing a final review playing with the code later. Please do not make any changes until then, as it would make me have to re-review some parts.

File: application.py
Comment (BattleMage0231) [reply application.py]: It seems to have changed after I forked the project. I have gotten rid of it.

File: application.py
Comment (BattleMage0231) [reply application.py]: This has been removed. See the previous comment from @ComputerGenius152

File: INSTALL.md
Comment (jdabtieu) [reply INSTALL.md:32]: the change is that it removes an extra space after `'point_value' integer NOT NULL DEFAULT`. This is for consistency.

**Comment by BattleMage0231** [comment]
Ok, please let me know when I should start making changes.

File: INSTALL.md
Comment (jdabtieu) [reply INSTALL.md:30]: Final review: looks ok

File: application.py
Comment (jdabtieu) [reply application.py]: Yep, I saw it.
Final review: resolved (outdated)

File: application.py
Comment (BattleMage0231) [reply application.py:708]: solved stores the problem ids of all problems solved by the user. It should have the same asymptotic complexity as before.

File: application.py
Comment (jdabtieu) [reply application.py:237]: Final review: looks good

File: helpers.py
Comment (BattleMage0231) [reply helpers.py:71]: It seems that many changes were made after I forked the repository yesterday.

File: application.py
Comment (jdabtieu) [reply application.py:411]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:423]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:429]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:446]: Initial review: see below
From the Python docs:
Warning Because naive datetime objects are treated by many datetime methods as local times, it is preferred to use aware datetimes to represent times in UTC. As such, the recommended way to create an object representing the current time in UTC is by calling datetime.now(timezone.utc).

File: application.py
Comment (jdabtieu) [inline application.py:365]: Final review: lines 374-404 need to be changed, this is still the old format, currently causing an Internal Server Error

File: application.py
Comment (jdabtieu) [reply application.py:471]: Final review: change
```py
    check1 = (len(check1) == 0)

    if check1:```
to
```py
    if len(check1) == 0:```

File: application.py
Comment (jdabtieu) [reply application.py:476]: Final review: Lines 484-485 will fail if the user doesn't yet exist in the database. Please add a check if user doesn't exists, insert them first

File: application.py
Comment (jdabtieu) [reply application.py:493]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:498]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:513]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:539]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:564]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:604]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:621]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:645]: Final review: looks ok

**Comment by ComputerGenius152** [comment]
@BattleMage0231 the database initialization commands in conftest.py needs to be updated with the commands in install.md. that should make the tests successful.

File: application.py
Comment (jdabtieu) [reply application.py:645]: Final review: Needs fix: `contest_id=cid` should be `contest_id=:cid`

File: application.py
Comment (jdabtieu) [reply application.py:680]: Cannot currently be verified

File: application.py
Comment (jdabtieu) [reply application.py:708]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:762]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:941]: Final review: resolving, in jdabtieu/CTFOJ:master this section is replaced by a simple regex search

File: application.py
Comment (jdabtieu) [reply application.py:972]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:1025]: Final review: looks good

File: application.py
Comment (jdabtieu) [reply application.py:1141]: Final review: looks good

File: application.py
Comment (jdabtieu) [inline application.py:1143]: Final review: No idea where this came from but it should not be here

File: templates/contest/draft_problems.html
Comment (jdabtieu) [reply templates/contest/draft_problems.html:18]: Final review: looks good

File: templates/problem/problems.html
Comment (jdabtieu) [reply templates/problem/problems.html:20]: Final review: looks good

File: templates/contest/contest.html
Comment (jdabtieu) [reply templates/contest/contest.html:36]: Final review: Can't verify, but 99.9% sure it's good

File: helpers.py
Comment (jdabtieu) [reply helpers.py:71]: Yes, there were approximately 80 commits in the previous ~30 hours. Anyways, I think you merged this code into application.py, but forgot to remove the old database queries. You can resolve this conversation if that's true.

**Review by jdabtieu — CHANGES_REQUESTED** [review]
See comments. You can start editing now.

**Comment by jdabtieu** [comment]
You can start editing now. @BattleMage0231

File: application.py
Comment (BattleMage0231) [reply application.py:1143]: This is when the contest is deleted (the contest tables are dropped). Before, the code was:

```python
db.execute("DROP TABLE :cid", cid=contest_id)	
db.execute("DROP TABLE :cidinfo", cidinfo=contest_id + "info")
```

Comment (jdabtieu) [reply application.py:1143]: no i'm talking about `os.remove('metadata')`

Comment (BattleMage0231) [reply application.py:1143]: I have no idea, it was in the repository when I forked it.

File: application.py
Comment (BattleMage0231) [reply application.py:476]: Should I insert the user into the contest database only if they get the flag correct or in all cases?

File: application.py
Comment (BattleMage0231) [reply application.py:680]: This basically indicates that everybody who solved the problem during the contest solved the problem. It selects everybody who did the problem and inserts into problem_solved that they solved it.

File: application.py
Comment (BattleMage0231) [reply application.py:476]: I only added them if they got the flag correct. Please comment if you want the other way or resolve the conversation if this is fine.

File: application.py
Comment (BattleMage0231) [reply application.py]: I have removed it.

File: application.py
Comment (jdabtieu) [reply application.py:484]: We should add in any case, as they made a submission. It doesn't matter if the submission is right or not, they made a submission, so they are part of the contest now.

File: application.py
Comment (jdabtieu) [reply application.py:688]: I couldn't verify since the contest stuff was broken, but I'll check it now.

**Review by jdabtieu — CHANGES_REQUESTED** [review]
Fix the insert into contest_users thing, once I finish checking the two pages I couldn't test earlier (contest stuff), I will update this review with more info

File: application.py
Comment (BattleMage0231) [reply application.py:484]: Fixed on the latest commit

**Comment by jdabtieu** [comment]
Ok, it turns out that using the python datetime.now(timezone.utc) isn't actually compatible with our app, and it's just better to use datetime.utcnow()
sorry about that, can you change it back? I've analyzed the other two things and found some egregious errors in exporting contest problems that I will now push. Other than that, once you revert all the datetimes, we should be ready to merge.

File: INSTALL.md
Comment (ComputerGenius152) [inline INSTALL.md:23]: this sparks joy

**Comment by BattleMage0231** [comment]
@jdabtieu The changes for datetime have been made.

File: tests/conftest.py
Comment (ComputerGenius152) [inline tests/conftest.py:18]: thank you for ixing the tests

**Review by ComputerGenius152 — APPROVED** [review]
looks good

**Comment by jdabtieu** [comment]
> @jdabtieu The changes for datetime have been made.

Ok, I'll do a final review. Once I approve it, if Anthony approves it too, we can merge it in.

**Review by jdabtieu — APPROVED** [review]
These changes look good to me. Everything seems to be working as intended.

@ComputerGenius152 please complete your review and if you approve it, please merge.

Nevermind, I didn't see that you approved 10 minutes ago. Merging into master now.

Thank you @BattleMage0231 for your contributions

**Pull request merged**
